airflow/models/serialized_dag.py

```diff
@@ -0,0 +1,155 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Serialzed DAG table in database."""
+
+import hashlib
+from typing import Any, Dict, List, Optional, TYPE_CHECKING
+from sqlalchemy import Column, Index, Integer, String, Text, and_
+from sqlalchemy.sql import exists
+
+from airflow.models.base import Base, ID_LEN
+from airflow.utils import db, timezone
+from airflow.utils.sqlalchemy import UtcDateTime
+
+
+if TYPE_CHECKING:
+    from airflow.dag.serialization.serialized_dag import SerializedDAG  # noqa: F401, E501; # pylint: disable=cyclic-import
+    from airflow.models import DAG  # noqa: F401; # pylint: disable=cyclic-import
+
+
+class SerializedDagModel(Base):
+    """A table for serialized DAGs.
+
+    serialized_dag table is a snapshot of DAG files synchronized by scheduler.
+    This feature is controlled by:
+        [core] dagcached = False: enable this feature
+        [core] dagcached_min_update_interval = 30 (s):
+            serialized DAGs are updated in DB when a file gets processed by scheduler,
+            to reduce DB write rate, there is a minimal interval of updating serialized DAGs.
+        [scheduler] dag_dir_list_interval = 300 (s):
+            interval of deleting serialized DAGs in DB when the files are deleted, suggest
+            to use a smaller interval such as 60
+
+    It is used by webserver to load dagbags when dagcached=True. Because reading from
+    database is lightweight compared to importing from files, it solves the webserver
+    scalability issue.
+    """
+    __tablename__ = 'serialized_dag'
+
+    dag_id = Column(String(ID_LEN), primary_key=True)
+    fileloc = Column(String(2000))
+    # The max length of fileloc exceeds the limit of indexing.
+    fileloc_hash = Column(Integer)
+    data = Column(Text)
+    last_updated = Column(UtcDateTime)
+
+    __table_args__ = (
+        Index('idx_fileloc_hash', fileloc_hash, unique=False),
+    )
+
+    def __init__(self, dag):
+        from airflow.dag.serialization import Serialization
+
+        self.dag_id = dag.dag_id
+        self.fileloc = dag.full_filepath
+        self.fileloc_hash = SerializedDagModel.dag_fileloc_hash(self.fileloc)
+        self.data = Serialization.to_json(dag)
```

Comment:
Either here, or inside to_json we should ensure that the JSON blob is valid - I want to minimize the chance of writing "odd"/invalid data in to our DB.

Reply:
Somewhere here I also want to add `version` field in to the JSON we store so that we can change this over time without having to "detect" the version based off what the stored data looks like.

Reply:
(This "version" is not the version/history of the dag, but the version/format of the serialization)

Reply:
> Either here, or inside to_json we should ensure that the JSON blob is valid - I want to minimize the chance of writing "odd"/invalid data in to our DB.

Done in https://github.com/apache/airflow/pull/5743/commits/977a2fe3fd244bc3f366a1228324f8b3c58f30ac . WDYT - is that OK?